File: lib/local-archive.ts
import fs from "node:fs";
import path from "node:path";
import localArchiveConfig from "@/config/local-archive.json";

const DATA_DIR = path.join(process.cwd(), "data");
const LOCAL_ARCHIVE_DIR = path.join(DATA_DIR, "local-archive");
const LOCAL_ARCHIVE_MANIFEST_PATH = path.join(LOCAL_ARCHIVE_DIR, "manifest.json");

export type LocalArchiveExtraEndpoint = {
  endpoint: string;
  id: string;
  label: string;
  method: string;
  source: string;
  target: string;
  timeoutSeconds: number;
};

export type LocalArchiveJobSummary = {
  archives: number;
  compressedBytes: number;
  label: string;
  lastArchivedAt: string | null;
  lastPath: string | null;
  lastScheduledAt: string | null;
  rawBytes: number;
  source: string;
};

export type LocalArchiveManifest = {
  generatedAt: string | null;
  jobs: Record<string, LocalArchiveJobSummary>;
  root: string;
  totals: {
    archives: number;
    compressedBytes: number;
    rawBytes: number;
  };
};

export type LocalArchiveStatus = {
  analysisRole: string;
  enabled: boolean;
  excludeJobIds: string[];
  excludedData: string[];
  extraEndpoints: LocalArchiveExtraEndpoint[];
  format: string;
  generatedAt: string;
  includeJobIds: string[];
  localStorageRole: string;
  manifest: LocalArchiveManifest;
  manifestExists: boolean;
  manifestPath: string;
  retentionYears: number;
  root: string;
  schedulerStatePath: string;
  supabaseRole: string;
  supabaseWindowYears: number;
  weatherDataPolicy: string;
};

type RawManifest = Partial<LocalArchiveManifest>;

function stringArray(value: unknown): string[] {
  return Array.isArray(value) ? value.map(String) : [];
}

function extraEndpoints(): LocalArchiveExtraEndpoint[] {
  return localArchiveConfig.extraEndpoints.map((endpoint) => ({
    endpoint: String(endpoint.endpoint),
    id: String(endpoint.id),
    label: String(endpoint.label),
    method: String(endpoint.method),
    source: String(endpoint.source),
    target: String(endpoint.target),
    timeoutSeconds: Number(endpoint.timeoutSeconds)
  }));
}

function emptyManifest(): LocalArchiveManifest {
  return {
    generatedAt: null,
    jobs: {},
    root: String(localArchiveConfig.root),
    totals: {
      archives: 0,
      compressedBytes: 0,
      rawBytes: 0
    }
  };
}

function normalizeManifest(value: RawManifest): LocalArchiveManifest {
  return {
    generatedAt: typeof value.generatedAt === "string" ? value.generatedAt : null,
    jobs: value.jobs && typeof value.jobs === "object" ? value.jobs as Record<string, LocalArchiveJobSummary> : {},
    root: typeof value.root === "string" ? value.root : String(localArchiveConfig.root),
    totals: {
      archives: Number(value.totals?.archives ?? 0),
      compressedBytes: Number(value.totals?.compressedBytes ?? 0),
      rawBytes: Number(value.totals?.rawBytes ?? 0)
    }
  };
}

export function getLocalArchiveStatus(): LocalArchiveStatus {
  let manifest = emptyManifest();
  let manifestExists = false;

  if (fs.existsSync(LOCAL_ARCHIVE_MANIFEST_PATH)) {
    manifestExists = true;
    try {
      manifest = normalizeManifest(JSON.parse(fs.readFileSync(LOCAL_ARCHIVE_MANIFEST_PATH, "utf-8")) as RawManifest);
    } catch {
      manifest = emptyManifest();
    }
  }

  return {
    analysisRole: String(localArchiveConfig.analysisRole),
    enabled: Boolean(localArchiveConfig.enabled),
    excludeJobIds: stringArray(localArchiveConfig.excludeJobIds),
    excludedData: stringArray(localArchiveConfig.excludedData),
    extraEndpoints: extraEndpoints(),
    format: String(localArchiveConfig.format),
    generatedAt: new Date().toISOString(),
    includeJobIds: stringArray(localArchiveConfig.includeJobIds),
    localStorageRole: String(localArchiveConfig.localStorageRole),
    manifest,
    manifestExists,
    manifestPath: String(localArchiveConfig.manifestPath),
    retentionYears: Number(localArchiveConfig.retentionYears),
    root: String(localArchiveConfig.root),
    schedulerStatePath: String(localArchiveConfig.schedulerStatePath),
    supabaseRole: String(localArchiveConfig.supabaseRole),
    supabaseWindowYears: Number(localArchiveConfig.supabaseWindowYears),
    weatherDataPolicy: String(localArchiveConfig.weatherDataPolicy)
  };
}

File: python/local_archive.py
from __future__ import annotations

import argparse
import json
from dataclasses import dataclass
from datetime import datetime
from zoneinfo import ZoneInfo

from data_scheduler import (
    call_job,
    iso,
    load_archive_config,
    load_config,
    load_jobs,
    save_local_archive,
    should_archive_job,
)

@dataclass(frozen=True)
class ArchiveEndpoint:
    endpoint: str
    id: str
    label: str
    method: str
    source: str
    target: str
    timeout_seconds: int


def load_extra_endpoints(archive_config: dict) -> list[ArchiveEndpoint]:
    endpoints = []
    for item in archive_config.get("extraEndpoints", []):
        endpoints.append(
            ArchiveEndpoint(
                endpoint=item["endpoint"],
                id=item["id"],
                label=item["label"],
                method=item.get("method", "GET"),
                source=item["source"],
                target=item["target"],
                timeout_seconds=int(item.get("timeoutSeconds", 120)),
            )
        )
    return endpoints


def main() -> None:
    parser = argparse.ArgumentParser(description="Archive current API payloads to local 10-year storage")
    parser.add_argument("--base-url", default=None)
    parser.add_argument("--region", default="kanto")
    parser.add_argument("--prefecture", default="tokyo")
    args = parser.parse_args()

    config = load_config()
    archive_config = load_archive_config()
    timezone = ZoneInfo(config["timezone"])
    base_url = args.base_url or config["defaultBaseUrl"]
    now = datetime.now(timezone)
    results = []

    archive_jobs = [
        job
        for job in load_jobs(config)
        if job.status == "active" and should_archive_job(job, archive_config)
    ]
    archive_jobs.extend(load_extra_endpoints(archive_config))

    for job in archive_jobs:
        if not should_archive_job(job, archive_config):
            continue

        try:
            payload = call_job(base_url, job, args.region, args.prefecture)
            archive = save_local_archive(archive_config, job, now, datetime.now(timezone), payload)
            results.append(
                {
                    "archive": archive,
                    "jobId": job.id,
                    "label": job.label,
                    "ok": True,
                    "source": job.source,
                }
            )
        except Exception as error:  # noqa: BLE001 - command-line report should keep going by job.
            results.append(
                {
                    "error": str(error),
                    "jobId": job.id,
                    "label": job.label,
                    "ok": False,
                    "source": job.source,
                }
            )

    print(
        json.dumps(
            {
                "generatedAt": iso(datetime.now(timezone)),
                "localArchive": {
                    "enabled": bool(archive_config.get("enabled", False)),
                    "excludedData": archive_config.get("excludedData", []),
                    "excludeJobIds": archive_config.get("excludeJobIds", []),
                    "extraEndpointIds": [endpoint.id for endpoint in load_extra_endpoints(archive_config)],
                    "includeJobIds": archive_config.get("includeJobIds", []),
                    "manifestPath": archive_config.get("manifestPath", "data/local-archive/manifest.json"),
                    "retentionYears": archive_config.get("retentionYears", 10),
                    "root": archive_config.get("root", "data/local-archive"),
                    "weatherDataPolicy": archive_config.get("weatherDataPolicy", ""),
                },
                "results": results,
                "timezone": config["timezone"],
            },
            ensure_ascii=False,
            indent=2,
        )
    )


if __name__ == "__main__":
    main()
